qlib/Logger.qm

```diff
@@ -0,0 +1,2731 @@
+        }
+
+        private auto serializeImpl(LoggerEvent event) {
+        }
+
+        public processEventImpl(int type, auto params) {
+        }
+    }
+
+    #! Implements pattern parsing and substitution
+    /**
+        The pattern defines fields to be substituted with real values
+
+        The pattern format is:
+
+        @code
+        %[<modifiers>] <word> [{<option>}]
+        modifiers := [0-9.-]*
+        word := [a-zA-Z]+
+        option := [^}]*
+        @endcode
+
+        When a word (i.e. longer than one char) key without curly brackets is not found then in the next step is resolved its first char,
+        e.g. \c "%parchive" the first step tries to resolve \c "parchive" key and if not resolved the second step \c "p" key.
+        To parse in key if single step delimit a word from the next character, use curly brackets, i.e. empty option, e.g. \c "%p{}archive".
+        If a key is not resolved then exception is raised.
+
+        The following table covers various modifiers and scenarios:
+
+        | !Format modifier | !left justify | !minimum width | !maximum width | !comment |
+        | %20c             | false         | 20             | none           | Left pad with spaces if the category name is less than 20 characters long |
+        | %-20c            | true          | 20             | none           | Right pad with spaces if the category name is less than 20 characters long |
+        | %.30c            | NA            | none           | 30             | Truncate from the beginning if the category name is longer than 30 characters |
+        | %20.30c          | false         | 20             | 30             | Left pad with spaces if the category name is shorter than 20 characters. However, if the category name is longer than 30 characters, then truncate from the beginning |
+        | %-20.30c         | true          | 20             | 30             | Right pad with spaces if the category name is shorter than 20 characters. However, if the category name is longer than 30 characters, then truncate from the beginning |
+    */
+    public class LoggerPattern {
+        public {
+            #! format escape character
+            const ESCAPE_CHAR = "%";
+        }
+
+        private:internal {
+            #! pattern
+            string m_pattern;
+        }
+
+        private {
+            #! parsed pattern
+            list m_parsedPattern;  # subclass may provide value e.g. for unit tests
+        }
+
+        #! Creates the object
+        /**
+            @param pattern the logger pattern
+            @throw LOGGER-ERROR when \a pattern is invalid
+        */
+        constructor(string pattern) {
+            # parse pattern
+            setPattern(pattern);
+        }
+
+        #! Sets the new pattern
+        /**
+            @param pattern the logger pattern
+            @throw LOGGER-ERROR when \a pattern is invalid
+        */
+        public setPattern(string pattern) {
```

Comment:
this method is really long with many one-char long variables, couldn't it be divided to smaller pieces with longer variable names?